From 5afa474a2a2cd0cadb592a54c66efd090c544a04 Mon Sep 17 00:00:00 2001
From: Yu Xia <yu.y.xia@intel.com>
Date: Thu, 31 May 2018 17:52:40 +0800
Subject: [PATCH 1954/2313] media: intel-ipu4p: imx477: add DOL resolutions

Message for Open Source:
Add 4056x2288 resolutions for DOL2/DOL3

Message for Internal:
[Issue/Feature] Refer to open source message
[Root Cause/Changes] Add sensor settings for DOL resolutions.

TrackedOn: #JKRLCSN-270
Change-Id: I144c890a423afe362511b11272de2a5d110585da
Signed-off-by: Yu Xia <yu.y.xia@intel.com>
Signed-off-by: Meng Wei <wei.meng@intel.com>
---
 .../media/i2c/crlmodule/crl_imx477_common_regs.h   |   2 +-
 .../crlmodule/crl_imx477_master_configuration.h    | 203 +++++++++++++++++++++
 2 files changed, 204 insertions(+), 1 deletion(-)

diff --git a/drivers/media/i2c/crlmodule/crl_imx477_common_regs.h b/drivers/media/i2c/crlmodule/crl_imx477_common_regs.h
index d83873a2..ecf20b9 100644
--- a/drivers/media/i2c/crlmodule/crl_imx477_common_regs.h
+++ b/drivers/media/i2c/crlmodule/crl_imx477_common_regs.h
@@ -10,7 +10,7 @@
 
 #include "crlmodule-sensor-ds.h"
 
-#define IMX477_CAPTURE_MODE_MAX	8
+#define IMX477_CAPTURE_MODE_MAX	10
 
 static struct crl_dynamic_register_access imx477_fll_regs[] = {
 	{
diff --git a/drivers/media/i2c/crlmodule/crl_imx477_master_configuration.h b/drivers/media/i2c/crlmodule/crl_imx477_master_configuration.h
index 16efa95..f28343c 100644
--- a/drivers/media/i2c/crlmodule/crl_imx477_master_configuration.h
+++ b/drivers/media/i2c/crlmodule/crl_imx477_master_configuration.h
@@ -979,6 +979,179 @@ static struct crl_register_write_rep imx477_656_512_19MHZ_master[] = {
 	{0x034F, CRL_REG_LEN_08BIT, 0x00}, /* Y output size [7:0]  */
 };
 
+static struct crl_register_write_rep imx477_4056_2288_19MHZ_DOL_2f_master[] = {
+	/* Frame Horizontal Clock Count */
+	{0x0342, CRL_REG_LEN_08BIT, 0x39}, /* Line length [15:8]  */
+	{0x0343, CRL_REG_LEN_08BIT, 0x14}, /* Line length [7:0]   */
+
+	/* Frame Vertical Clock Count */
+	{0x0340, CRL_REG_LEN_08BIT, 0x20}, /* Frame length [15:8] */
+	{0x0341, CRL_REG_LEN_08BIT, 0x11}, /* Frame length [7:0]  */
+
+	/* Visible Size */
+	/* (0,376) to (4055, 2664) */
+	{0x0344, CRL_REG_LEN_08BIT, 0x00}, /* Analog cropping start X [12:8] */
+	{0x0345, CRL_REG_LEN_08BIT, 0x00}, /* Analog cropping start X [7:0]  */
+	{0x0346, CRL_REG_LEN_08BIT, 0x01}, /* Analog cropping start Y [12:8] */
+	{0x0347, CRL_REG_LEN_08BIT, 0x78}, /* Analog cropping start Y [7:0]  */
+	{0x0348, CRL_REG_LEN_08BIT, 0x0F}, /* Analog cropping end X [12:8]   */
+	{0x0349, CRL_REG_LEN_08BIT, 0xD7}, /* Analog cropping end X [7:0]    */
+	{0x034A, CRL_REG_LEN_08BIT, 0x0A}, /* Analog cropping end Y [12:8]   */
+	{0x034B, CRL_REG_LEN_08BIT, 0x68}, /* Analog cropping end Y [7:0]    */
+
+	/* Mode Setting */
+	{0x00E3, CRL_REG_LEN_08BIT, 0x01}, /* DOL-HDR enabled */
+	{0x00E4, CRL_REG_LEN_08BIT, 0x01}, /* DOL Mode: DOL2 */
+	{0x00FC, CRL_REG_LEN_08BIT, 0x0A}, /* The output data fmt for CSI: RAW10 */
+	{0x00FD, CRL_REG_LEN_08BIT, 0x0A}, /* The output data fmt for CSI: RAW10 */
+	{0x3E10, CRL_REG_LEN_08BIT, 0x01}, /* VC ID of DOL 2nd frame */
+
+	{0x0220, CRL_REG_LEN_08BIT, 0x00}, /* Undocumented */
+	{0x0221, CRL_REG_LEN_08BIT, 0x11}, /* Undocumented */
+	{0x0381, CRL_REG_LEN_08BIT, 0x01}, /* Num of pixels skipped, even -> odd */
+	{0x0383, CRL_REG_LEN_08BIT, 0x01}, /* Num of pixels skipped, odd -> even */
+	{0x0385, CRL_REG_LEN_08BIT, 0x01}, /* Num of lines skipped, even -> odd  */
+	{0x0387, CRL_REG_LEN_08BIT, 0x01}, /* Num of lines skipped, odd -> even  */
+	{0x0900, CRL_REG_LEN_08BIT, 0x00}, /* Binning mode: Disable */
+	{0x0901, CRL_REG_LEN_08BIT, 0x11}, /* Binning Type for Horizontal */
+	{0x0902, CRL_REG_LEN_08BIT, 0x02}, /* Binning Type for Vertical   */
+	{0x3140, CRL_REG_LEN_08BIT, 0x02}, /* Undocumented */
+	{0x3C00, CRL_REG_LEN_08BIT, 0x00}, /* Undocumented */
+	{0x3C01, CRL_REG_LEN_08BIT, 0x03}, /* Undocumented */
+	{0x3C02, CRL_REG_LEN_08BIT, 0xDC}, /* Undocumented */
+	{0x3F0D, CRL_REG_LEN_08BIT, 0x00}, /* AD converter: 10 bit */
+	{0x5748, CRL_REG_LEN_08BIT, 0x07}, /* Undocumented */
+	{0x5749, CRL_REG_LEN_08BIT, 0xFF}, /* Undocumented */
+	{0x574A, CRL_REG_LEN_08BIT, 0x00}, /* Undocumented */
+	{0x574B, CRL_REG_LEN_08BIT, 0x00}, /* Undocumented */
+	{0x7B75, CRL_REG_LEN_08BIT, 0x0E}, /* Undocumented */
+	{0x7B76, CRL_REG_LEN_08BIT, 0x09}, /* Undocumented */
+	{0x7B77, CRL_REG_LEN_08BIT, 0x0C}, /* Undocumented */
+	{0x7B78, CRL_REG_LEN_08BIT, 0x06}, /* Undocumented */
+	{0x7B79, CRL_REG_LEN_08BIT, 0x3B}, /* Undocumented */
+	{0x7B53, CRL_REG_LEN_08BIT, 0x01}, /* Undocumented */
+	{0x9369, CRL_REG_LEN_08BIT, 0x5A}, /* Undocumented */
+	{0x936B, CRL_REG_LEN_08BIT, 0x55}, /* Undocumented */
+	{0x936D, CRL_REG_LEN_08BIT, 0x28}, /* Undocumented */
+	{0x9304, CRL_REG_LEN_08BIT, 0x03}, /* Undocumented */
+	{0x9305, CRL_REG_LEN_08BIT, 0x00}, /* Undocumented */
+	{0x9E9A, CRL_REG_LEN_08BIT, 0x2F}, /* Undocumented */
+	{0x9E9B, CRL_REG_LEN_08BIT, 0x2F}, /* Undocumented */
+	{0x9E9C, CRL_REG_LEN_08BIT, 0x2F}, /* Undocumented */
+	{0x9E9D, CRL_REG_LEN_08BIT, 0x00}, /* Undocumented */
+	{0x9E9E, CRL_REG_LEN_08BIT, 0x00}, /* Undocumented */
+	{0x9E9F, CRL_REG_LEN_08BIT, 0x00}, /* Undocumented */
+	{0xA2A9, CRL_REG_LEN_08BIT, 0x60}, /* Undocumented */
+	{0xA2B7, CRL_REG_LEN_08BIT, 0x00}, /* Undocumented */
+
+	/* Digital Crop & Scaling */
+	{0x0401, CRL_REG_LEN_08BIT, 0x00}, /* Scaling mode: No Scaling     */
+	{0x0404, CRL_REG_LEN_08BIT, 0x00}, /* Down Scaling Factor M [8]    */
+	{0x0405, CRL_REG_LEN_08BIT, 0x10}, /* Down Scaling Factor M [7:0]  */
+	{0x0408, CRL_REG_LEN_08BIT, 0x00}, /* Crop Offset from X [12:8]    */
+	{0x0409, CRL_REG_LEN_08BIT, 0x00}, /* Crop Offset from X [7:0]     */
+	{0x040A, CRL_REG_LEN_08BIT, 0x00}, /* Crop Offset from Y [12:8]    */
+	{0x040B, CRL_REG_LEN_08BIT, 0x00}, /* Crop Offset from Y [7:0]     */
+	{0x040C, CRL_REG_LEN_08BIT, 0x0F}, /* Width after cropping [12:8]  */
+	{0x040D, CRL_REG_LEN_08BIT, 0xD8}, /* Width after cropping [7:0]   */
+	{0x040E, CRL_REG_LEN_08BIT, 0x08}, /* Height after cropping [12:8] */
+	{0x040F, CRL_REG_LEN_08BIT, 0xF0}, /* Height after cropping [7:0]  */
+
+	/* Output Crop */
+	{0x034C, CRL_REG_LEN_08BIT, 0x0F}, /* X output size [12:8] */
+	{0x034D, CRL_REG_LEN_08BIT, 0xD8}, /* X output size [7:0]  */
+	{0x034E, CRL_REG_LEN_08BIT, 0x08}, /* Y output size [12:8] */
+	{0x034F, CRL_REG_LEN_08BIT, 0xF0}, /* Y output size [7:0]  */
+};
+
+static struct crl_register_write_rep imx477_4056_2288_19MHZ_DOL_3f_master[] = {
+	/* Frame Horizontal Clock Count */
+	{0x0342, CRL_REG_LEN_08BIT, 0x39}, /* Line length [15:8]  */
+	{0x0343, CRL_REG_LEN_08BIT, 0x14}, /* Line length [7:0]   */
+
+	/* Frame Vertical Clock Count */
+	{0x0340, CRL_REG_LEN_08BIT, 0x20}, /* Frame length [15:8] */
+	{0x0341, CRL_REG_LEN_08BIT, 0x11}, /* Frame length [7:0]  */
+
+	/* Visible Size */
+	/* (0,376) to (4055, 2664) */
+	{0x0344, CRL_REG_LEN_08BIT, 0x00}, /* Analog cropping start X [12:8] */
+	{0x0345, CRL_REG_LEN_08BIT, 0x00}, /* Analog cropping start X [7:0]  */
+	{0x0346, CRL_REG_LEN_08BIT, 0x01}, /* Analog cropping start Y [12:8] */
+	{0x0347, CRL_REG_LEN_08BIT, 0x78}, /* Analog cropping start Y [7:0]  */
+	{0x0348, CRL_REG_LEN_08BIT, 0x0F}, /* Analog cropping end X [12:8]   */
+	{0x0349, CRL_REG_LEN_08BIT, 0xD7}, /* Analog cropping end X [7:0]    */
+	{0x034A, CRL_REG_LEN_08BIT, 0x0A}, /* Analog cropping end Y [12:8]   */
+	{0x034B, CRL_REG_LEN_08BIT, 0x68}, /* Analog cropping end Y [7:0]    */
+
+	/* Mode Setting */
+	{0x00E3, CRL_REG_LEN_08BIT, 0x01}, /* DOL-HDR Disable */
+	{0x00E4, CRL_REG_LEN_08BIT, 0x02}, /* DOL Mode: DOL3 */
+	{0x00FC, CRL_REG_LEN_08BIT, 0x0A}, /* The output data fmt for CSI: RAW10 */
+	{0x00FD, CRL_REG_LEN_08BIT, 0x0A}, /* The output data fmt for CSI: RAW10 */
+	{0x00FE, CRL_REG_LEN_08BIT, 0x0A}, /* The output data fmt for CSI: RAW10 */
+	{0x00FF, CRL_REG_LEN_08BIT, 0x0A}, /* The output data fmt for CSI: RAW10 */
+	{0x3E10, CRL_REG_LEN_08BIT, 0x01}, /* VC ID of DOL 2nd frame */
+	{0x3E11, CRL_REG_LEN_08BIT, 0x02}, /* VC ID of DOL 3rd frame */
+
+	{0x0220, CRL_REG_LEN_08BIT, 0x00}, /* Undocumented */
+	{0x0221, CRL_REG_LEN_08BIT, 0x11}, /* Undocumented */
+	{0x0381, CRL_REG_LEN_08BIT, 0x01}, /* Num of pixels skipped, even -> odd */
+	{0x0383, CRL_REG_LEN_08BIT, 0x01}, /* Num of pixels skipped, odd -> even */
+	{0x0385, CRL_REG_LEN_08BIT, 0x01}, /* Num of lines skipped, even -> odd  */
+	{0x0387, CRL_REG_LEN_08BIT, 0x01}, /* Num of lines skipped, odd -> even  */
+	{0x0900, CRL_REG_LEN_08BIT, 0x00}, /* Binning mode: Disable */
+	{0x0901, CRL_REG_LEN_08BIT, 0x11}, /* Binning Type for Horizontal */
+	{0x0902, CRL_REG_LEN_08BIT, 0x02}, /* Binning Type for Vertical   */
+	{0x3140, CRL_REG_LEN_08BIT, 0x02}, /* Undocumented */
+	{0x3C00, CRL_REG_LEN_08BIT, 0x00}, /* Undocumented */
+	{0x3C01, CRL_REG_LEN_08BIT, 0x03}, /* Undocumented */
+	{0x3C02, CRL_REG_LEN_08BIT, 0xDC}, /* Undocumented */
+	{0x3F0D, CRL_REG_LEN_08BIT, 0x00}, /* AD converter: 10 bit */
+	{0x5748, CRL_REG_LEN_08BIT, 0x07}, /* Undocumented */
+	{0x5749, CRL_REG_LEN_08BIT, 0xFF}, /* Undocumented */
+	{0x574A, CRL_REG_LEN_08BIT, 0x00}, /* Undocumented */
+	{0x574B, CRL_REG_LEN_08BIT, 0x00}, /* Undocumented */
+	{0x7B75, CRL_REG_LEN_08BIT, 0x0E}, /* Undocumented */
+	{0x7B76, CRL_REG_LEN_08BIT, 0x09}, /* Undocumented */
+	{0x7B77, CRL_REG_LEN_08BIT, 0x0C}, /* Undocumented */
+	{0x7B78, CRL_REG_LEN_08BIT, 0x06}, /* Undocumented */
+	{0x7B79, CRL_REG_LEN_08BIT, 0x3B}, /* Undocumented */
+	{0x7B53, CRL_REG_LEN_08BIT, 0x01}, /* Undocumented */
+	{0x9369, CRL_REG_LEN_08BIT, 0x5A}, /* Undocumented */
+	{0x936B, CRL_REG_LEN_08BIT, 0x55}, /* Undocumented */
+	{0x936D, CRL_REG_LEN_08BIT, 0x28}, /* Undocumented */
+	{0x9304, CRL_REG_LEN_08BIT, 0x03}, /* Undocumented */
+	{0x9305, CRL_REG_LEN_08BIT, 0x00}, /* Undocumented */
+	{0x9E9A, CRL_REG_LEN_08BIT, 0x2F}, /* Undocumented */
+	{0x9E9B, CRL_REG_LEN_08BIT, 0x2F}, /* Undocumented */
+	{0x9E9C, CRL_REG_LEN_08BIT, 0x2F}, /* Undocumented */
+	{0x9E9D, CRL_REG_LEN_08BIT, 0x00}, /* Undocumented */
+	{0x9E9E, CRL_REG_LEN_08BIT, 0x00}, /* Undocumented */
+	{0x9E9F, CRL_REG_LEN_08BIT, 0x00}, /* Undocumented */
+	{0xA2A9, CRL_REG_LEN_08BIT, 0x60}, /* Undocumented */
+	{0xA2B7, CRL_REG_LEN_08BIT, 0x00}, /* Undocumented */
+
+	/* Digital Crop & Scaling */
+	{0x0401, CRL_REG_LEN_08BIT, 0x00}, /* Scaling mode: No Scaling     */
+	{0x0404, CRL_REG_LEN_08BIT, 0x00}, /* Down Scaling Factor M [8]    */
+	{0x0405, CRL_REG_LEN_08BIT, 0x10}, /* Down Scaling Factor M [7:0]  */
+	{0x0408, CRL_REG_LEN_08BIT, 0x00}, /* Crop Offset from X [12:8]    */
+	{0x0409, CRL_REG_LEN_08BIT, 0x00}, /* Crop Offset from X [7:0]     */
+	{0x040A, CRL_REG_LEN_08BIT, 0x00}, /* Crop Offset from Y [12:8]    */
+	{0x040B, CRL_REG_LEN_08BIT, 0x00}, /* Crop Offset from Y [7:0]     */
+	{0x040C, CRL_REG_LEN_08BIT, 0x0F}, /* Width after cropping [12:8]  */
+	{0x040D, CRL_REG_LEN_08BIT, 0xD8}, /* Width after cropping [7:0]   */
+	{0x040E, CRL_REG_LEN_08BIT, 0x08}, /* Height after cropping [12:8] */
+	{0x040F, CRL_REG_LEN_08BIT, 0xF0}, /* Height after cropping [7:0]  */
+
+	/* Output Crop */
+	{0x034C, CRL_REG_LEN_08BIT, 0x0F}, /* X output size [12:8] */
+	{0x034D, CRL_REG_LEN_08BIT, 0xD8}, /* X output size [7:0]  */
+	{0x034E, CRL_REG_LEN_08BIT, 0x08}, /* Y output size [12:8] */
+	{0x034F, CRL_REG_LEN_08BIT, 0xF0}, /* Y output size [7:0]  */
+};
+
 static struct crl_mode_rep imx477_modes_master[] = {
 	{
 		.sd_rects_items = ARRAY_SIZE(imx477_4056_3040_rects),
@@ -1100,6 +1273,36 @@ static struct crl_mode_rep imx477_modes_master[] = {
 		.mode_regs_items = ARRAY_SIZE(imx477_656_512_19MHZ_master),
 		.mode_regs = imx477_656_512_19MHZ_master,
 	},
+	{
+		.sd_rects_items = ARRAY_SIZE(imx477_4056_2288_rects),
+		.sd_rects = imx477_4056_2288_rects,
+		.binn_hor = 1,
+		.binn_vert = 1,
+		.scale_m = 1,
+		.width = 4056,
+		.height = 2288,
+		.min_llp = 14612,
+		.min_fll = 8209,
+		.comp_items = 0,
+		.ctrl_data = 0,
+		.mode_regs_items = ARRAY_SIZE(imx477_4056_2288_19MHZ_DOL_2f_master),
+		.mode_regs = imx477_4056_2288_19MHZ_DOL_2f_master,
+	},
+	{
+		.sd_rects_items = ARRAY_SIZE(imx477_4056_2288_rects),
+		.sd_rects = imx477_4056_2288_rects,
+		.binn_hor = 1,
+		.binn_vert = 1,
+		.scale_m = 1,
+		.width = 4056,
+		.height = 2288,
+		.min_llp = 14612,
+		.min_fll = 8209,
+		.comp_items = 0,
+		.ctrl_data = 0,
+		.mode_regs_items = ARRAY_SIZE(imx477_4056_2288_19MHZ_DOL_3f_master),
+		.mode_regs = imx477_4056_2288_19MHZ_DOL_3f_master,
+	},
 };
 
 static struct crl_flip_data imx477_flip_configurations_master[] = {
-- 
2.7.4

